engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/BuiltinRootNode.java

```diff
@@ -52,4 +69,174 @@ public void setTypeName(QualifiedName typeName) {
    */
   @Override
   public abstract String getName();
+
+  protected static final class ArgContext {
+    private TruffleObject returnValue;
+    private EnsoHashMap warnings;
+
+    public ArgContext() {}
+
+    public TruffleObject getReturnValue() {
+      return returnValue;
+    }
+
+    public boolean hasWarnings() {
+      return this.warnings != null;
+    }
+
+    private void addWarnings(
+        VirtualFrame frame, HashMapInsertAllNode insertNode, EnsoHashMap newWarnings) {
+      if (this.warnings == null) {
+        this.warnings = newWarnings;
+      } else {
+        int maxWarnings = EnsoContext.get(insertNode).getWarningsLimit();
+        this.warnings = insertNode.executeInsertAll(frame, this.warnings, newWarnings, maxWarnings);
+      }
+    }
+  }
+
+  protected abstract static class ArgNode extends Node {
+    private static final byte IS_SELF = 0x01;
+    private static final byte IS_ARRAY = 0x02;
+    private static final byte REQUIRES_CAST = 0x04;
+    private static final byte CHECK_ERRORS = 0x08;
+    private static final byte CHECK_PANIC_SENTINEL = 0x10;
+    private static final byte CHECK_WARNINGS = 0x20;
+    private final byte flags;
+    @CompilerDirectives.CompilationFinal private Type ensoType;
+
+    @Child private WarningsLibrary warnings;
+    @Child private AppendWarningNode appendWarningNode;
+    @Child private HashMapInsertAllNode mapInsertAllNode;
+
+    ArgNode(byte flags) {
+      this.flags = flags;
+      if (is(CHECK_WARNINGS)) {
+        this.warnings = WarningsLibrary.getFactory().createDispatched(5);
+      }
+    }
+
+    private boolean is(byte what) {
+      return (flags & what) != 0;
+    }
+
+    @SuppressWarnings("unchecked")
+    public final <T> T processArgument(
+        VirtualFrame frame, Class<T> type, Object value, ArgContext context) {
+      assert value != null;
+      if (is(CHECK_ERRORS) && value instanceof DataflowError err) {
+        context.returnValue = err;
+        return null;
+      }
+      if (is(CHECK_PANIC_SENTINEL) && value instanceof PanicSentinel sentinel) {
+        throw sentinel.getPanic();
+      }
+      if (warnings != null) {
+        if (warnings.hasWarnings(value)) {
+          if (mapInsertAllNode == null) {
+            CompilerDirectives.transferToInterpreterAndInvalidate();
+            this.mapInsertAllNode = HashMapInsertAllNode.build();
+          }
+          if (this.mapInsertAllNode != null) {
```

Comment:
This `if` can be removed

Reply:
I don't think it can! `mapInsertAllNode` is a compilation constant and this avoids the code in `if` branch to be PEed. Only when the `mapInsertAllNode` is non-null, the branch is known to be "taken" - replaces the `BranchProfile` that used to be there before.

Reply:
Until we find first `value` that is `instanceof WithWarnings` then the code for PE compilation should be just:
```java
if (value instanceof WithWarnings) {
  transferToInterpreterAndInvalidate();
```
nothing else is visible for the compiler. Once the value of `mapInsertAllNode` changes, we want the code to be:
```java
if (value instanceof WithWarnings) {
      context.addWarnings(frame, mapInsertAllNode, warnings.getWarnings(value, false));
      value = warnings.removeWarnings(value);
}
```
which it is. But it would be so even without the `if (this.mapInsertAllNode != null)` check...

> This `if` can be removed

I have to agree. The `if` should be removed.